knative/deployer.go

```diff
@@ -7,73 +7,89 @@ import (
 	"os"
 	"strings"
 
-	"github.com/boson-project/faas"
-	"github.com/boson-project/faas/k8s"
 	commands "knative.dev/client/pkg/kn/commands"
 	"knative.dev/client/pkg/kn/core"
+
+	"github.com/boson-project/faas"
+	"github.com/boson-project/faas/k8s"
 )
 
 // TODO: Use knative.dev/serving/pkg/client/clientset/versioned/typed/serving/v1
 // NewForConfig gives you the client, and then you can do
 // client.Services("ns").Get("name")
 
-func NewDeployer() *Deployer {
-	return &Deployer{Namespace: faas.DefaultNamespace}
-}
-
 type Deployer struct {
+	// Namespace with which to override that set on the default configuration (such as the ~/.kube/config).
+	// If left blank, deployment will commence to the configured namespace.
 	Namespace string
-	Verbose   bool
+	// Verbose logging enablement flag.
+	Verbose bool
 }
 
-func (deployer *Deployer) Deploy(name, image string) (address string, err error) {
+func NewDeployer() *Deployer {
+	return &Deployer{}
+}
 
-	project, err := k8s.ToSubdomain(name)
+func (d *Deployer) Deploy(f faas.Function) (err error) {
+	// k8s does not support service names with dots.  so encode it such that
+	// www.my-domain,com -> www-my--domain-com
```

Comment:
Personally, I don't like the double dashes where `my-domain` becomes `my--domain`

Reply:
I most certainly do not either.  Unfortunately that's a restriction placed on us by k8s, which requires the positively draconian [RFC 1035](https://tools.ietf.org/html/rfc1035) label:
> labels must follow the rules for ARPANET host names.  They must
start with a letter, end with a letter or digit, and have as interior
characters only letters, digits, and hyphen.

So restrictive, the only option I could think of was to overload hyphen with itself or make it ambiguous :/

Reply:
But isn't that still ambiguous if somebody uses two consecutive hyphens?